### DISCUSSION TITLE
Portenta H7 File Corruption Issues

### DISCUSSION THREAD

**Discussion by AngeloFatica** [micropython/micropython#13192] [STM32 / Pyboard]
Hello,

Using the standard micropython build from the repository for the Portenta I am having some issues with files. The compilation of the firmware works well and the dfu uploads fine, but the file system reading and writing is where I am having issues. I am uploading around 100KB in files to the file system using thonny to do so. Most of the time, whether I upload all at once or one at a time, I get a syntax error at random points in my files. When going to investigate the file, I try to open it and thonny sends me an error screen. There is different errors presented usually, some say issues in thonnys back end, some say it cannot recognize the encoding of the file anymore, and some say that thonny needs to be restarted altogether. Ive had similar issues in different IDE's as well like Arduino Lab and using rshell. I do not want to freeze my code just yet as I am still developing on the board. I have not had this issue ever with the STM32 F405 chip or any others, just with the portenta. I am using the external flash for the file system, which has about 16MB. Any help with this is appreciated.

**Comment by andrewleech** [comment]
Could you try copying the files with mpremote, it's the officially supported tool here: https://docs.micropython.org/en/latest/reference/mpremote.html

**Reply by AngeloFatica** [reply]
Can this be used in CLI for MAC OS? Seems to be tailored for a Linux system. Tried running python3 -m mpremote and got:

`Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/angelofatica/mpremote/__main__.py", line 4, in <module>
    from mpremote import main
  File "/Users/angelofatica/mpremote/main.py", line 25, in <module>
    from .commands import (
  File "/Users/angelofatica/mpremote/commands.py", line 5, in <module>
    import serial.tools.list_ports
ModuleNotFoundError: No module named 'serial'`

Any help with this is appreciated.

**Reply by AngeloFatica** [reply]
And for the record I do have serial python package installed ^^

**Reply by andrewleech** [reply]
It does work on macos, should be exactly the same.
Was this after a `pip install mpremote`? The dependencies should be handled automatically.

**Reply by AngeloFatica** [reply]
yes it was. This is what I am getting when just trying to run 'mpremote':
`zsh: command not found: mpremote`

When trying to run it as a python3 script I got the error mentioned above.

**Reply by andrewleech** [reply]
Might be an issue with conflicting / separate copies of python. Maybe try:
`python3 -m pip install mpremote`

**Reply by AngeloFatica** [reply]
This did the trick. Got into the portenta with this, but am experiencing new issues with the board. I detailed them below. Thanks for the help with mpremote!

**Comment by AngeloFatica** [comment]
I would also like to add, when building micropython with threading enabled in 'mpconfigport.h' it bricks my module as in all the RGB lights turn on and I cannot access it. This does not happen when I set the threading bit to 0 and rebuild and deploy onto the same board. Not sure why the portenta micropython build does this as I have had a successful thread build for it before?

This also happens sometimes when the board is reset. I have had this happen on two different boards with different builds. Even with the v1.21.0 build.

**Reply by andrewleech** [reply]
The H7 is a relatively new chip / port, perhaps the threading support isn't very stable. In my experience, threads aren't used that much in micropython so not as well tested on different chips / hardware. `ayncio` is generally promoted as the preferred concurrency pattern.

**Reply by AngeloFatica** [reply]
I think it is a file system issue as reflashing micropython dfu (and erasing flash) at 0x8040000 doesn't fix the problem, but erasing the flash at 0x8020000 which is the file system space I believe does fix the issue. The issue arises once again when files are attempted to be flashed into the file system. Maybe a linker memory mapping issue?

**Comment by AngeloFatica** [comment]
bump

I think it is a file system issue as reflashing micropython dfu (and erasing flash) at 0x8040000 doesn't fix the problem, but erasing the flash at 0x8020000 which is the file system space I believe does fix the issue. The issue arises once again when files are attempted to be flashed into the file system. Maybe a linker memory mapping issue?

**Comment by andrewleech** [comment]
Possibly related to https://github.com/micropython/micropython/issues/13063

**Reply by andrewleech** [reply]
Might also be related / fixed by https://github.com/micropython/micropython/pull/13141

**Comment by AngeloFatica** [comment]
I would like to point out that I did not compile the mboot bootloader for the portenta before flashing with micropython. I completely missed the warning message in my build logs. Not sure if this is a complete fix but is probably the cause, I wonder if the flash is being written starting in the wrong spot causing all my errors.

**Edit: I ended up bricking both of my portentas playing around with them, the PMIC chip is no longer programmed at boot so I cannot get power delivery to the rest of the boards anymore 🤦‍♂️.
